### PR TITLE
[FIX] product: variant sale price is wrong

### PR DESCRIPTION
**Pull request by momegahed**
If applied, this commit will fix the following bug by using the price
with the additional cost added instead of only the price

Steps to reproduce:

1- install sales
2- create a product p with variant pv
3- set an extra value for pv
4- go to Products>Product Variants
5- open the pv in form view
6- the Sales Price is the price of p not pv
(which should be p + extra value)
7- also the tax string is calculated on p not pv

Bug:

```list_price``` is being used

Fix:

replace ```list_price``` with  ```lst_price```

OPW-2774199

### CODE REVIEW

**Review by adwid — CHANGES_REQUESTED** [review]
Considering how `lst_price` is computed (`= list_price + extra_prices`) + the fact that `lst_price` is the one used on the tree view of `product.product`:
https://github.com/odoo/odoo/blob/d3ec94f1c79ecb76acb8de027bf47b0f0b7a3b0f/addons/product/views/product_views.xml#L336
Your changes in the XML file seem legit

However, I don't understand why there is an issue with the computation of `tax_string` ? Note that the addition of non-stored fields should be avoided as much as possible. Also, the implementation of `product_product._compute_tax_string` is strange. `product.product` is a child of `product.template`, so the method overrides the original one while its body calls the original one

Also, could you be more precise in the steps-to-reproduce ? I'm not sure it is that clear that the user has to set an extra-price on a specific variant attribute (I have had to read the ticket again :innocent:)

**Comment by momegahed** [comment]
> Considering how `lst_price` is computed (`= list_price + extra_prices`) + the fact that `lst_price` is the one used on the tree view of `product.product`:
> 
> https://github.com/odoo/odoo/blob/d3ec94f1c79ecb76acb8de027bf47b0f0b7a3b0f/addons/product/views/product_views.xml#L336
> 
> 
> Your changes in the XML file seem legit
> However, I don't understand why there is an issue with the computation of `tax_string` ? Note that the addition of non-stored fields should be avoided as much as possible. Also, the implementation of `product_product._compute_tax_string` is strange. `product.product` is a child of `product.template`, so the method overrides the original one while its body calls the original one
> 
> Also, could you be more precise in the steps-to-reproduce ? I'm not sure it is that clear that the user has to set an extra-price on a specific variant attribute (I have had to read the ticket again 😇)

the inheritance between `product.product` and `product.template` is [delegation inheritance](https://www.odoo.com/documentation/15.0/developer/reference/backend/orm.html#delegation) not traditional inheritance

thus the method override wouldn't work

okay to the second point

**Comment by adwid** [comment]
My bad for the inheritance
I am still not convinced by the changes, I didn't find a use of `tax_string` anywhere but on a form of a product template:
https://github.com/odoo/odoo/blob/1f69445f0a7ed02030f3f16f126e34b0d507ed80/addons/account/views/product_view.xml#L76
but maybe I'm missing something ?
@agr-odoo Could you check this ?

**Comment by momegahed** [comment]
> My bad for the inheritance I am still not convinced by the changes, I didn't find a use of `tax_string` anywhere but on a form of a product template:
> 
> https://github.com/odoo/odoo/blob/1f69445f0a7ed02030f3f16f126e34b0d507ed80/addons/account/views/product_view.xml#L76
> 
> 
> but maybe I'm missing something ?
> @agr-odoo Could you check this ?

`product.product` form view interits from `product.template` and the same field is needed there albeit it needs correction in the calculation thus the need for the changes

**Review by agr-odoo — CHANGES_REQUESTED** [review]
I'm ok with the idea. This needs to be fixed as price and tax string of the variant product shows the template product data

On the actual implementation it would be better to have also the opinion of @Feyensv

**Comment by Feyensv** [comment]
I also don't understand the need of the tax_string code and I don't see any mention of it in the source ticket.
If nobody needs it, don't add it, it's not like something is wrong without the tax explanation (furthermore, existing database won't have the change anyway).

For the `lst_price` change, could you investigate why it works fine in previous versions, and maybe link the commit/PR that caused the problem ? (probably #75862 but I cannot say for sure)

File: addons/product/views/product_views.xml [hunk 1]
```diff
@@ -365,9 +365,10 @@
                 </xpath>
                 <field name="list_price" position="attributes">
                    <attribute name="attrs">{'readonly': [('product_variant_count', '&gt;', 1)]}</attribute>
+                   <attribute name="invisible">1</attribute>
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/views/product_views.xml:368]: Doesn't `attrs` attribute expect domains ?
I think a separate invisible attribute is cleaner and clearer here :)

**Comment by agr-odoo** [comment]
> I also don't understand the need of the tax_string code and I don't see any mention of it in the source ticket. If nobody needs it, don't add it, it's not like something is wrong without the tax explanation (furthermore, existing database won't have the change anyway).

The tax explanation needs to be fixed (or hided) as it uses the values of the template product
Example: 
Template product have a sales price 750.00 with 15% tax incl 
    -> tax string is "(= $ 862.50 Incl. Taxes)"
Variant product add +50.00
    -> tax string before patch is the same "(= $ 862.50 Incl. Taxes)" (see screenshot)
    -> tax string after patch is "(= $ 920.00 Incl. Taxes)"
screenshot_005

**Comment by Feyensv** [comment]
Okay, makes sense now, but there really should be some explanation about it in the PR/commit, which isn't the case currently ...

File: addons/product/views/product_views.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/product/views/product_views.xml:368]: > Doesn't `attrs` attribute expect domains ?

One could put a true leaf/domain (optimized to avoid queries etc.)
https://github.com/odoo/odoo/blob/561fee585643cdf3ea3af2ef548b2d5fd1d86a1c/odoo/osv/expression.py#L176

But I agree that the invisible node attribute is better

File: addons/account/models/product.py [hunk 2]
```diff
@@ -63,24 +63,28 @@ def get_product_accounts(self, fiscal_pos=None):
...
-            if currency.compare_amounts(excluded, record.list_price):
-                joined.append(_('%s Excl. Taxes', format_amount(self.env, excluded, currency)))
-            if joined:
-                record.tax_string = f"(= {', '.join(joined)})"
-            else:
-                record.tax_string = " "
-
+            record.tax_string = self._construct_tax_string(record, record.list_price)
+    @api.model
+    def _construct_tax_string(record, price):
```
Comment (william-andre) [inline addons/account/models/product.py:68]:
```suggestion

    def _construct_tax_string(self, price):
```

This is not `api.model` at all since you are using `record`, which should be named `self` to respect python's naming convention, and preceded by an empty line again to follow python's PEP8 conventions

File: addons/account/models/product.py [hunk 3]
```diff
@@ -63,24 +63,28 @@ def get_product_accounts(self, fiscal_pos=None):
...
+        if joined:
+            tax_string = f"(= {', '.join(joined)})"
+        else:
+            tax_string = " "
+        return tax_string
 
 class ProductProduct(models.Model):
     _inherit = "product.product"
 
+    tax_string = fields.Char(compute='_compute_tax_string')
```
Comment (william-andre) [inline addons/account/models/product.py:87]: add an empty line after
Is it even needed? I don't see anything that is changed from the inherited field declaration. Redefining the computed function implementation should be enough.

Comment (momegahed) [reply addons/account/models/product.py:87]: @william-andre 
I tried that but it didn't work, the product.template implementation is one that is always called. I though it is because of the way that [delegation inheritance (https://www.odoo.com/documentation/15.0/developer/reference/backend/orm.html#delegation) works

Comment (william-andre) [reply addons/account/models/product.py:88]: Anyways, add an empty line, and adapt the dependencies.

I'll test it afterwards :)

File: addons/account/models/product.py [hunk 4]
```diff
@@ -156,3 +162,8 @@ def _get_tax_included_unit_price(self, company, currency, document_date, documen
             product_price_unit = product_currency._convert(product_price_unit, currency, company, document_date)
 
         return product_price_unit
+
+    def _compute_tax_string(self):
+        for record in self:
+            record.tax_string = record.product_tmpl_id._construct_tax_string(record, record.lst_price)
```
Comment (william-andre) [inline addons/account/models/product.py:168]: This tracebacks, and it is missing dependencies

**Review by william-andre — COMMENTED** [review]
Please mark the PR as ready so that I can approve it when you are finished with the last changes :)

File: addons/account/models/product.py (hunk 4 above)
Comment (momegahed) [reply addons/account/models/product.py]: > Please mark the PR as ready so that I can approve it when you are finished with the last changes :)

well, my fear is that because now `_construct_tax_string` is an instance method and the inheritance is delegation (so polymorphism won't work), it will be very confusing later when extending this. for now it is not an issue because the only attribute used to calculate the tax string is the price but imagine wanting to add other variables that might differ between product.product and product.template.

 tl;dr
`self` in `_construct_tax_string` is always `product_tmpl_id` regardless of where it was called from

Comment (momegahed) [reply addons/account/models/product.py]: also why the style check fails with C0305, I added one line at the end

https://github.com/odoo/odoo/pull/87248/files#diff-6193fe8fb1a4b291ce7c82ef509968e9b381312909738d621adda02b2177807fR169

Comment (william-andre) [reply addons/account/models/product.py]: > also why the style check fails with C0305, I added one line at the end

You have 2 end of line char instead of one, leading in one blank/useless line

> `self` in `_construct_tax_string` is always `product_tmpl_id` regardless of where it was called from

How is that answering any of my remarks?
And why is that an issue?

Comment (momegahed) [reply addons/account/models/product.py]: I did fix the errors, I don't know what dependencies it misses.

this was another point.

the issue is that the intuitive behavior when someone sees this implementation will be that this is a normal case of polymorphism and can thus be extended further which is not the case

File: addons/account/models/product.py [hunk 5]
```diff
@@ -156,3 +162,8 @@ def _get_tax_included_unit_price(self, company, currency, document_date, documen
             product_price_unit = product_currency._convert(product_price_unit, currency, company, document_date)
 
         return product_price_unit
+
+    def _compute_tax_string(self):
+        for record in self:
+            record.tax_string = record.product_tmpl_id._construct_tax_string(record.lst_price)
```
Comment (william-andre) [inline addons/account/models/product.py:168]: This depends on `lst_price`, `product_tmpl_id` and `taxes_id`

**Review by william-andre — APPROVED** [review]
@robodoo r+
